File: backend/schemas/project.py
from pydantic import BaseModel
from typing import Optional
from datetime import datetime


class ProjectBase(BaseModel):
    name: str
    description: Optional[str] = ""
    region: Optional[str] = "武汉市"
    budget_period: Optional[str] = ""
    notes: Optional[str] = ""
    status: Optional[str] = "进行中"


class ProjectCreate(ProjectBase):
    pass


class ProjectUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    region: Optional[str] = None
    budget_period: Optional[str] = None
    notes: Optional[str] = None
    status: Optional[str] = None


class ProjectResponse(ProjectBase):
    id: int
    created_by: str
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    quota_count: Optional[int] = 0
    total_cost: Optional[float] = 0.0

    class Config:
        from_attributes = True


class ProjectQuotaItem(BaseModel):
    """项目中的单条定额"""
    quota_id: str
    project_name: Optional[str] = None
    section: Optional[str] = None
    unit: Optional[str] = None
    quantity: float
    total_cost: float
    labor_fee: Optional[float] = None
    material_fee: Optional[float] = None
    machinery_fee: Optional[float] = None
    management_fee: Optional[float] = None
    tax: Optional[float] = None


class ProjectDetailResponse(ProjectResponse):
    """项目详情（含定额列表）"""
    items: list[ProjectQuotaItem] = []


class ProjectQuotaAdd(BaseModel):
    """向项目添加定额"""
    quota_id: str
    quantity: float = 1.0

File: frontend/src/api/project.ts
import api from './auth'

export interface Project {
  id: number
  name: string
  description: string
  region: string
  budget_period: string
  notes: string
  status: string
  created_by: string
  created_at: string
  updated_at: string | null
  quota_count: number
  total_cost: number
}

export interface ProjectQuotaItem {
  quota_id: string
  project_name: string | null
  section: string | null
  unit: string | null
  quantity: number
  total_cost: number
  labor_fee: number | null
  material_fee: number | null
  machinery_fee: number | null
  management_fee: number | null
  tax: number | null
}

export interface ProjectDetail extends Project {
  items: ProjectQuotaItem[]
}

export interface ProjectCreate {
  name: string
  description?: string
  region?: string
  budget_period?: string
  notes?: string
  status?: string
}

export interface ProjectUpdate {
  name?: string
  description?: string
  region?: string
  budget_period?: string
  notes?: string
  status?: string
}

// 获取项目列表
export async function listProjects(params?: { skip?: number; limit?: number; status?: string }): Promise<Project[]> {
  const res = await api.get<Project[]>('/api/v1/projects/', { params })
  return res.data
}

// 获取项目详情
export async function getProject(projectId: number): Promise<ProjectDetail> {
  const res = await api.get<ProjectDetail>(`/api/v1/projects/${projectId}`)
  return res.data
}

// 创建项目
export async function createProject(data: ProjectCreate): Promise<Project> {
  const res = await api.post<Project>('/api/v1/projects/', data)
  return res.data
}

// 更新项目
export async function updateProject(projectId: number, data: ProjectUpdate): Promise<Project> {
  const res = await api.patch<Project>(`/api/v1/projects/${projectId}`, data)
  return res.data
}

// 删除项目
export async function deleteProject(projectId: number): Promise<void> {
  await api.delete(`/api/v1/projects/${projectId}`)
}

// 添加定额到项目
export async function addQuotaToProject(projectId: number, quotaId: string, quantity: number = 1): Promise<ProjectQuotaItem> {
  const res = await api.post<ProjectQuotaItem>(`/api/v1/projects/${projectId}/quotas`, {
    quota_id: quotaId,
    quantity,
  })
  return res.data
}

// 从项目移除定额
export async function removeQuotaFromProject(projectId: number, quotaId: string): Promise<void> {
  await api.delete(`/api/v1/projects/${projectId}/quotas/${quotaId}`)
}

// 更新项目中定额数量
export async function updateQuotaQuantity(projectId: number, quotaId: string, quantity: number): Promise<void> {
  await api.patch(`/api/v1/projects/${projectId}/quotas/${quotaId}`, null, {
    params: { quantity },
  })
}
